File: prover_rclcpp/src/ros2cli_965.cpp
#include <rclcpp/rclcpp.hpp>

int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);

  auto node = rclcpp::Node::make_shared("param_demo_cpp");

  rcl_interfaces::msg::ParameterDescriptor descriptor;
  descriptor.type = rclcpp::ParameterType::PARAMETER_STRING;

#if 0
  // this needs to be templated, so rclcpp does not allow user to set parameter without default nor type.
  //node->declare_parameter(  // cannot compile
  node->declare_parameter<std::string>( // rclcpp::exceptions::UninitializedStaticallyTypedParameterException
    "param_2_cpp",
    descriptor);
#else
  // this works as designed
  descriptor.dynamic_typing = true;
  descriptor.type = rclcpp::ParameterType::PARAMETER_NOT_SET;
  node->declare_parameter(
    "param_2_cpp",
    rclcpp::ParameterValue{},
    descriptor);
#endif

  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(node);
  executor.spin();
  rclcpp::shutdown();

  return 0;
}

File: prover_rclpy/src/ros2cli_965.py
import rclpy
import rclpy.node
from rcl_interfaces.msg import ParameterDescriptor, ParameterType


class ParamDemo(rclpy.node.Node):
    def __init__(self):
        super().__init__("param_demo_py")

        #self.declare_parameter( # this should generate ValueError but humble
        #    "param_2_py",
        #    descriptor=ParameterDescriptor(type=ParameterType.PARAMETER_STRING),
        #)

        self.declare_parameter(
            "param_2_py",
            #value=ParameterValue(),
            descriptor=ParameterDescriptor(type=ParameterType.PARAMETER_NOT_SET, dynamic_typing=True),
        )

def main():
    rclpy.init()
    node = ParamDemo()
    rclpy.spin(node)


if __name__ == "__main__":
    main()